### DISCUSSION TITLE
Can someone please explain _prep_out_transaction() in cdc_device.c?

### DISCUSSION THREAD

**Discussion by clhenry** [hathach/tinyusb#1832] [Q&A] [answered]
I'm specifically referring to `TU_VERIFY(available >= sizeof(p_cdc->epout_buf));` and `if ( available >= sizeof(p_cdc->epout_buf) )`.

I'm running into an issue that I think might be a bug but it might also be a knowledge/understanding gap.

I'll be outlining my thought train so please feel free to correct me if I go off the rails.

So here's my scenario:

I'm evaluating TinyUSB as a replacement for TI's MSP430 USB stack currently in use in one of our products.
On the journey I encountered `tud_cdc_rx_cb()` being called by `cdcd_xfer_cb()`.
According to `if (tud_cdc_rx_cb && !tu_fifo_empty(&p_cdc->rx_ff) ) tud_cdc_rx_cb(itf);` the callback should execute on each receipt of data as long as the receive FIFO (configured for 64 bytes) isn't empty.

What I actually observed was that the callback only fired on the first byte received. No events were generated on subsequent writes by the host and sniffing the bus showed that the endpoint kept NAKing while moserial was attempting to send the rest of the data. The only solution was to read all received data and/or flush the receive FIFO and then the endpoint would ACK again then go back to NAKing while the host was trying to send more data.

Investigating this led me to `_prep_out_transaction()` and the lines of code I outlined at the top of this discussion.

The solution that worked for me was to change the `available >= sizeof(p_cdc->epout_buf)`s to ` available > 0`s. This change had the effect of allowing `usbd_edpt_xfer()` to be called which cleared the NAKs on my out endpoint. It also caused `tud_cdc_rx_cb()` to get called on the receipt of each byte up to the point where the receive FIFO filled up.

So, is this a bug? As is the checks in question would never pass and the out endpoint would never be available to receive anymore data without intervention.

**Comment by HiFiPhile** [comment, marked as answer]
> The only solution was to read all received data and/or flush the receive FIFO and then the endpoint would ACK again then go back to NAKing while the host was trying to send more data.

It's by design to ensure all data are processed before the host can send more.
`available >= sizeof(p_cdc->epout_buf)` is needed because we don't know how much data host will send at next transfer, so there is a need to have at least `sizeof(p_cdc->epout_buf)` free space before scheduling a new transfer.
If you want the host to send more data before you do the processing, you can increase `CFG_TUD_CDC_RX_BUFSIZE`.

**Comment by clhenry** [comment]
Thanks @HiFiPhile. You definitely cleared things up for me.

For anyone else that might run into the same issue the problem boiled down to a misconfiguration as highlighted by @HiFiPhile.

`CFG_TUD_CDC_EP_BUFSIZE` was set to 64 which is the same size as my non-0 endpoints and `CFG_TUD_CDC_RX_BUFSIZE` was also set to 64.

The host was sending one byte at a time and upon receipt of the first byte the CDC receive buffer only had space for 63 more bytes.

With the endpoint being capable of receiving up to 64 bytes on the next transfer the stack refused to clear the endpoint NAK since **all** the received data (potentially 64 bytes) would have nowhere to go.
